WordPress/gutenberg
# Use browser parser APIs for block-editor style transformations.

## What?
This PR uses the browser API for parsing and processing the frontend styles.
Where the browser API is yet insufficient, few, small, feature-focused and well-tested libraries are used.

## Why?
Because a custom CSS parser is currently used (that is not developed as part of a dedicated parser project), many issues with style parsing and processing occur, often crashing the post-processing completely, resulting in no editor styles applied at all.

Note: A previous attempt that uses the [`css-tree`](https://www.npmjs.com/package/css-tree) parser has unacceptable size requirements (https://github.com/WordPress/gutenberg/pull/25514).

## How?
This PR implements the CSS post-processing by leveraging the currently supported browser APIs, supplemented by small, feature-focused libraries, where browser support is yet insufficient, allows to increase robustness, performance and size efficiency.

## Testing Instructions
1. Use the Gutenberg plugin with this PR (branch) applied.
2. Add frontend styles or other styles as editor stylesheet that are known to cause issues with the custom CSS parser (e.g. https://github.com/WordPress/gutenberg/issues/21569; https://github.com/WordPress/gutenberg/issues/19930 and from other issues).
3. Open the Gutenberg editor, 
note that the editor styles are correctly post-processed and applied, resulting in correctly wrapped selectors and rewritten URLs.

### Testing Instructions for Keyboard
(UX is and should not be changed by this PR)

## Screenshots or screencast 
(Not applicable)
PoC of style post-processing using the browser API: https://codepen.io/strarsis/pen/MWmMQgy

## Additional notes
I had issues with installing the `npm` dependencies (Ubuntu `22.04` LTS on WSL `2` on Windows `10`), albeit the correct `node` (by `.nvmrc`) and `npm` (by `package.json` constraints) versions were used, `npm install` failed with `error cb() never called!`.
The solution was self-upgrading `npm`, but not to the latest available version of it (which would be incompatible with the required constraints by `package.json`), rather to its version that comes with that `node` version:
- `nvm install`
- `nvm use`
- (`npm install` would result in `error cb() never called!` on a new `node` installation (although `npm` is of version `6.14.18`).
- `npm install -g npm@6.14.18` # self-upgrade `npm` to its latest version in the constraints required by `package.json`
- `npm install` # installation works now.
